src/Features/Core/Portable/AddRequiredParentheses/AbstractAddRequiredParenthesesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,101 @@
+{
+    internal abstract class AbstractAddRequiredParenthesesDiagnosticAnalyzer<
+        TLanguageKindEnum,
+        TBinaryExpressionSyntax>
+        : AbstractCodeStyleDiagnosticAnalyzer
+        where TLanguageKindEnum : struct
+        where TBinaryExpressionSyntax : SyntaxNode
+    {
+        protected enum OperationKind
+        {
+            Arithmetic,
+            Logical
+        }
+
+        protected AbstractAddRequiredParenthesesDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.AddRequiredParenthesesDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Add_parentheses_for_clarity), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources.Parentheses_should_be_added_for_clarity), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public sealed override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, GetSyntaxNodeKinds());
+
+        protected abstract ImmutableArray<TLanguageKindEnum> GetSyntaxNodeKinds();
+        protected abstract bool CanAddParentheses(
+            TBinaryExpressionSyntax binaryExpression, SemanticModel semanticModel, out OperationKind operationKind);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var syntaxTree = context.SemanticModel.SyntaxTree;
+            var cancellationTokan = context.CancellationToken;
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationTokan).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var binaryExpression = (TBinaryExpressionSyntax)context.Node;
+
+            var arithmeticPreference = optionSet.GetOption(CodeStyleOptions.ArithmeticOperationParentheses, binaryExpression.Language);
+            var logicalPreference = optionSet.GetOption(CodeStyleOptions.LogicalOperationParentheses, binaryExpression.Language);
+            if (arithmeticPreference.Value != ArithmeticOperationParenthesesPreference.RequireForPrecedenceClarity &&
+                logicalPreference.Value != LogicalOperationParenthesesPreference.RequireForPrecedenceClarity)
+            {
+                // User wants to keep all parentheses, even if unnecessary.
+                return;
+            }
+
+            if (!CanAddParentheses(binaryExpression, context.SemanticModel,
+                    out var operationKind))
+            {
+                return;
+            }
+
+            if (!ShouldAddParentheses())
```

Comment:
`ShouldAddParenthesesForOperationKind`? It's confusing because you're not _passing_ operationKind (such that I could infer the actual meaning of this line), but you're just referencing it from the local function. I'd rather this not be a local function, but I might be biased.